File: HW3/cipher.py
def encrypt(message, N, e):
    result = []
    for i in range(0, len(message), 2):
        chunk = ord(message[i]) << 8
        try:
            chunk |= ord(message[i+1])
        except IndexError:
            pass
        chunk = pow(chunk, e, N)
        result.append(chunk)
    return result


if __name__ == '__main__':
    with open('plain.txt', 'r') as f:
        message = f.read().strip()
    with open('public_key.txt', 'r') as f:
        N = int(f.readline().strip())
        e = int(f.readline().strip())
    with open('secret.txt', 'w') as f:
        for chunk in encrypt(message, N, e):
            f.write(str(chunk))
            f.write('\n')

File: HW3/cipher.cpp
#include <cstdint>
#include <fstream>
#include <string>
#include "pow.h"

template<class T, class = typename std::enable_if<std::is_integral<T>::value>::type>
void encrypt(std::ofstream& ofs, std::string& message, T N, T e) {
    auto end = message.length() - 1;
    T temp = 0;
    for(unsigned i=0; i<end; i+=2) {
        temp = (message[i] << 8) | message[i+1];
        ofs << '\n' << pow(temp, e, N);
    }
    if (end%2 == 0) { // end+1 is odd => end is even
        temp = message[end] << 8;
        ofs << '\n' << pow(temp, e, N);
    }
    ofs.flush();
}

int main() {
    std::ifstream msg("plain.txt"), pub("public_key.txt");
    std::ofstream ofs("secret.txt");
    std::string message;
    msg.seekg(0, std::ios::end);
    message.resize(msg.tellg());
    msg.seekg(0, std::ios::beg);
    msg.read(&message[0], message.length());
    uint64_t N, e;
    pub >> N >> e;
    encrypt(ofs, message, N, e);
    return 0;
}